crypto/fipsmodule/kdf/sskdf.c

```diff
@@ -0,0 +1,336 @@
+  if (!ctx || !ctx->data) {
+    return 0;
+  }
+  sskdf_variant_hmac_ctx *variant_ctx = (sskdf_variant_hmac_ctx *)ctx->data;
+  if (!variant_ctx) {
+    return 0;
+  }
+  return HMAC_size(variant_ctx->hmac_ctx);
+}
+
+static int sskdf_variant_hmac_compute(sskdf_variant_ctx *ctx, uint8_t *out,
+                                      size_t out_len,
+                                      const uint8_t counter[SSKDF_COUNTER_SIZE],
+                                      const uint8_t *secret, size_t secret_len,
+                                      const uint8_t *info, size_t info_len) {
+  if (!ctx || !ctx->data || !out || !counter || !secret) {
+    return 0;
+  }
+
+  sskdf_variant_hmac_ctx *variant_ctx = (sskdf_variant_hmac_ctx *)ctx->data;
+
+  if (!variant_ctx->hmac_ctx) {
+    return 0;
+  }
+
+  uint32_t written;
+
+  if (!HMAC_Init_ex(variant_ctx->hmac_ctx, NULL, 0, NULL, NULL) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, &counter[0], SSKDF_COUNTER_SIZE) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, secret, secret_len) ||
+      !HMAC_Update(variant_ctx->hmac_ctx, info, info_len) ||
+      !HMAC_Final(variant_ctx->hmac_ctx, out, &written) || out_len != written) {
+    return 0;
+  }
+
+  return 1;
+}
+
+
+DEFINE_METHOD_FUNCTION(sskdf_variant, sskdf_variant_digest) {
+  memset(out, 0, sizeof(sskdf_variant));
+  out->output_size = sskdf_variant_digest_output_size;
+  out->compute = sskdf_variant_digest_compute;
+}
+
+
+DEFINE_METHOD_FUNCTION(sskdf_variant, sskdf_variant_hmac) {
+  memset(out, 0, sizeof(sskdf_variant));
+  out->output_size = sskdf_variant_hmac_output_size;
+  out->compute = sskdf_variant_hmac_compute;
+}
+
+static int SSKDF(const sskdf_variant *variant, sskdf_variant_ctx *ctx,
+                 uint8_t *out_key, size_t out_len, const uint8_t *secret,
+                 size_t secret_len, const uint8_t *info, size_t info_len) {
+  int ret = 0;
+
+  if (!ctx || !variant) {
+    return 0;
+  }
+
+  if (!out_key || out_len == 0 || !secret || secret_len == 0) {
+    goto err;
+  }
+
+  // output_size is the length in bytes of output of the SSKDF variant auxilary
+  // function (EVP_DigestFinal or HMAC_Final)
+  size_t output_size = variant->output_size(ctx);
+  if (output_size == 0) {
+    goto err;
+  }
+  assert(output_size <= EVP_MAX_MD_SIZE);
+
+  // NIST.SP.800-56Cr2:
+  // 1. If L > 0, then set reps = ceil(L / H_outputBits); otherwise, output an
+  // error indicator and exit this process without performing the remaining
+  // actions
+  uint64_t n = (out_len + output_size - 1) / output_size;
+
+  // NIST.SP.800-56Cr2:
+  // 2. If reps > (2^32 - 1), then output an error indicator and exit this
+  // process without performing the remaining actions
+  if (n > UINT32_MAX) {
+    goto err;
+  }
+
+  // Validate that the sum of the fixed info lengths won't overflow a size_t
+  // which will very based on 32-bit or 64-bit platforms. We will further clamp
+  // this value below.
+  if (info_len > SIZE_MAX - 4 - secret_len) {
```

Comment:
it's not clear to me why we have this and the next check? why not test only with `UINT32_MAX`?

Reply:
The original intention of this check (and the original variation which broke out the check per variable) is to ensure that summing the `4 + secret_len + info_len` doesn't overflow size_t. Just checking that the sum is less then UINT32_MAX would not catch that conidtion. If the lengths are sufficiently large then the total length can wrap around and it would go uncaught.

> Unsigned integer arithmetic is always performed modulo 2n where n is the number of bits in that particular integer. E.g. for unsigned int, adding one to UINT_MAX gives ​0​, and subtracting one from ​0​ gives UINT_MAX. 
See https://en.cppreference.com/w/c/language/operator_arithmetic

I doubted myself when you initially made this comment, and collapsed it down to this one statement, but on review again I think validating this is important for correctly validating the later condition required in step 4 of the specification.